File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-3.cpp
class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int ROWS = ssize(heights), COLS = ssize(heights[0]);
        set<pair<int, int>> pacific;
        set<pair<int, int>> atlantic;

        auto dfs = [&](this auto& self, int r, int c, set<pair<int, int>>& visited, int prev_height) -> void {
            if (min(r, c) < 0 ||
                r == ROWS ||
                c == COLS ||
                visited.contains({r, c}) ||
                heights[r][c] < prev_height) return;
            
            visited.insert({r, c});
            self(r - 1, c, visited, heights[r][c]);
            self(r, c + 1, visited, heights[r][c]);
            self(r + 1, c, visited, heights[r][c]);
            self(r, c - 1, visited, heights[r][c]);
        };

        for (int r = 0; r < ROWS; r++) {
            dfs(r, 0, pacific, heights[r][0]);
            dfs(r, COLS - 1, atlantic, heights[r][COLS - 1]);
        }

        for (int c = 0; c < COLS; c++) {
            dfs(0, c, pacific, heights[0][c]);
            dfs(ROWS - 1, c, atlantic, heights[ROWS - 1][c]);
        }

        vector<vector<int>> result;
        for (const auto& coords : pacific) {
            if (atlantic.contains(coords)) {
                result.push_back({coords.first, coords.second});
            }
        }
        return result;
    }
};

File: Data Structures & Algorithms/surrounded-regions/submission-3.cs
public class Solution {
    public void Solve(char[][] board) {
        int ROWS = board.Length, COLS = board[0].Length;

        void Dfs(int r, int c) {
            if (Math.Min(r, c) < 0 ||
                r == ROWS ||
                c == COLS ||
                board[r][c] != 'O') return;
            
            board[r][c] = 'T';
            Dfs(r - 1, c);
            Dfs(r, c + 1);
            Dfs(r + 1, c);
            Dfs(r, c - 1);
        }

        int[] borderRows = { 0, ROWS - 1 };
        int[] borderCols = { 0, COLS - 1 };

        for (int r = 0; r < ROWS; r++) {
            foreach (int c in borderCols) {
                if (board[r][c] == 'O') Dfs(r, c);
            }
        }

        for (int c = 0; c < COLS; c++) {
            foreach (int r in borderRows) {
                if (board[r][c] == 'O') Dfs(r, c);
            }
        }

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (board[r][c] == 'T') board[r][c] = 'O';
                else if (board[r][c] == 'O') board[r][c] = 'X';
            }
        }
    }
}
